doc/developer/design/20210630_linearizability.md

```diff
@@ -0,0 +1,84 @@
+
+This is limited to transactions (or single statements) in the SQL layer where materialize is allowed to choose the timestamp for an operation.
+Statements can use `AS OF` to specify the timestamp of an operation.
+
+Since dataflow provides serializability, the addition of linearizability would also provide strict serializability over the SQL layer (mostly, see Catalog discussion below).
+
+[Jepsen]: http://jepsen.io/consistency/models/linearizable
+
+## Non-Goals
+
+Linearizability across the entire system.
+This may come in a later document.
+
+Verifying the necessary behaviors of other parts of the system.
+
+## Description
+
+The coordinator acts as if dataflow returns results at a timestamp only if it will always return the same result at that timestamp, even after restart.
+We are aware that this is not the case today, but document this necessary behavior to draw clear lines of responsibility.
+For example, we have volatile sources whose data changes after restart,
+and non-volatile sources that may have different timestamp bindings after restart because they return results before timestamps are durably persisted.
+
+To provide linearizability, when a SQL transaction (which can be a single statement) needs to generate a timestamp (either at commit time for INSERT or the first SELECT or TAIL in a transaction),
+that timestamp must be >= any other timestamp generated by any other transaction timestamp that is generated after it in real time.
+This will ensure that should some SQL result be observed by some user,
+that user can tell some other user about its results,
+and the second user is guaranteed to observe those results (unless they've changed in the meantime).
+
+However, we cannot have a single timestamp number that we guarantee always goes up, because timestamps do not always have the same meaning.
+We have solved the different meaning problem with timelines, which allow joining different sources only if their timestamp number means the same thing.
+For linearizability, we need to have a timestamp number for each timeline.
+
+Add a `HashMap<Timeline, Timestamp>` to track the highest timestamp for a given timeline.
+`sequence_peek`, `sequence_tail`, and `sequence_end_transaction` are the three places timestamps are generated for transactions.
+These three functions will call some new helper function that will either advance or return the most recent timestamp in the HashMap.
+
+This linearizability system can thus be viewed as a layer above the current timestamp determination functions, timelines, and time domains, since it specifically only cares about timestamps that users initiated.
+
+### System Clock
+
+The system clock (calls to `SystemTime::now()`) contributes to timestamps for realtime sources, table reads, and table writes.
+It may decrease (due to, say, some time synchronization program like NTP setting the clock).
+Linearizability is preserved in this case with the above `HashMap`.
```

Comment:
Does that above hashmap actually inform realtime sources? I had thought that they were out of the control of the coordinator.

Reply:
The hashmap does not inform realtime sources of anything. The sentence here is trying to convey that the hashmap will prevent timestamp selection for queries (table writes or reads from tables or rt sources) from ever decreasing, even if the system clock decreased.

Reply:
I see. So, it contributes to the times for the reads of realtime sources and tables, and the writes of realtime tables?

Reply:
Yes. Is there something wrong or unclear here that I need to correct?

Reply:
> contributes to timestamps for realtime sources,

The main thing that realtime sources do is produce timestamps *for data*, and that is not what is meant here afaict.

Reply:
More specifically, would you mind changing 
> contributes to timestamps for realtime sources, table reads, and table writes.
to 
> contributes to timestamps for reads from realtime sources and tables, and writes to tables.
or the equivalent. Just about being clear that the coordinator's clock does not have a role in producing timestamps for "writes" to realtime sources (i.e. the timestamp assigned to untimestamped data).